src/components/app/index.tsx

```diff
@@ -2,17 +2,58 @@ import { h, Component } from 'preact';
 import { bind, bitmapToImageData } from '../../lib/util';
 import * as style from './style.scss';
 import Output from '../output';
+import Options from '../options';
 
-import {MozJpegEncoder} from '../../lib/codec-wrappers/mozjpeg-enc';
+import { Encoder } from '../../lib/codec-wrappers/codec';
+import { MozJpegEncoder } from '../../lib/codec-wrappers/mozjpeg-enc';
+
+export type ImageType = 'original' | 'jpeg';
+
+export type CodecOptions = any;
+
+type Encoders = {
+  [type: string]: new () => Encoder
+};
+
+const EncoderNames = {
+  original: 'Original Image',
+  jpeg: (options: CodecOptions) => `JPEG ${options.quality || ''}`
+};
+
+const AllEncoders: Encoders = {
+  jpeg: MozJpegEncoder
+};
 
 type Props = {};
 
 type State = {
-  img?: ImageBitmap
+  sourceImg?: ImageBitmap,
+  sourceData?: ImageData,
+  leftImg?: ImageBitmap,
+  rightImg?: ImageBitmap,
+  loading: boolean
+  leftType: ImageType,
+  rightType: ImageType,
+  leftOptions: CodecOptions,
+  rightOptions: CodecOptions
 };
 
 export default class App extends Component<Props, State> {
-  state: State = {};
+  state: State = {
+    loading: false,
+    leftType: 'original',
+    rightType: 'jpeg',
+    leftOptions: {},
+    rightOptions: {}
+  };
+
+  optionsUpdateTimer?: NodeJS.Timer | number | null;
```

Comment:
- Don’t think you need `null` if you are using `?`
- Why `NodeJS.Timer`? 
- What does this prop stand for?

Reply:
It only works with this combination. I believe it's because we're combining the Node and web global typescript defs - setTimeout() returns a NodeJS.Timer, but cancelTimeout expects an int.